File: 2023/1/src/Day01.kt
import java.io.File
import java.net.URL
import java.util.regex.Pattern

// Constants
val INPUT_FILE = "../input.txt"

fun readInputData(inputFile: String): List<String> {
    val cwd = File("").absolutePath
    val file = File("$cwd/$inputFile")
    return if (file.exists()) {
        println("Using $inputFile for input data.")
        file.readLines()
    } else {
        println("Input file $inputFile not found. Fetching data from AoC website.")
        // Exit
        emptyList<String>().also { System.exit(1) }
    }
}

  /**
   * Part 1
   * 1. Find all digits in the string
   * 2. Add the first and last digits. If there is only one digit, add it twice
   * 3. Repeat for all lines
   * 
   * @param input
   * @return sum of all digits
   */
fun part1(input: List<String>): Int {
    var total = 0
    for (x in input.indices) {
        val digits = input[x].filter { it.isDigit() }
        if (digits.isNotEmpty()) {
            total += (digits.first().toString() + digits.last().toString()).toInt()
        }
    }
    return total
}

  /**
   * Part 2
   * 1. Find all digits in the string. Convert words to digits
   * 2. Add the first and last digits. If there is only one digit, add it twice
   * 3. Repeat for all lines
   * 
   * @param lines
   * @return sum of all digits
   */
fun part2(lines: List<String>): Int {
    var total = 0
    val wordToNum = mapOf("one" to 1, "two" to 2, "three" to 3, "four" to 4, "five" to 5, "six" to 6, "seven" to 7, "eight" to 8, "nine" to 9)
    val sortedKeys = wordToNum.keys.sortedBy { it.length }
    val pattern = Regex(sortedKeys.joinToString("|") + "|\\d+")
    for (line in lines) {
        println("Line: $line")
        val words = mutableListOf<String>()
        var i = 0
        while (i < line.length) {
            val match = pattern.find(line.substring(i))
            if (match != null) {
                words.add(match.value)
                i += match.value.length
            } else {
                i += 1
            }
        }
        println("Words: $words")
        val number = words.map { word -> wordToNum[word]?.toString() ?: word }.joinToString("").toInt()
        val digits = number.toString().map { it.toString().toInt() }
        if (digits.isNotEmpty()) {
            println("Digits: $digits")
            val left = digits.first()
            val right = digits.last()
            println("Left: $left, Right: $right")
            total += (left + right).toInt()
            print("Total: $total")
        }
    }
    return total
}

fun alternative(lines: List<String>): Int {
    val wordToNum = mapOf("one" to 1, "two" to 2, "three" to 3, "four" to 4, "five" to 5, "six" to 6, "seven" to 7, "eight" to 8, "nine" to 9)
    var total = 0
    val splitter = "(?=(" + wordToNum.keys.joinToString("|") + "|\\d+))"
    val pattern = Regex(splitter)
    for (line in lines) {
        val matches = pattern.findAll(line)
        println("Line: $line")

        val digits = matches.mapNotNull { match -> 
            if (match.value.isNotEmpty()) {
                if (match.value.all { it.isDigit() }) {
                    match.value.toInt()
                } else {
                    wordToNum[match.value]
                }
            } else {
                null
            }
        }.toList()
        if (digits.isNotEmpty()) {
            println("Digits: $digits")
            val left = digits.first()
            val right = digits.last()
            println("Left: $left, Right: $right")
            total += (left + right).toInt()
        }
    }
    return total
}

fun main() {
    val test_data1 = """
        1abc2
        pqr3stu8vwx
        a1b2c3d4e5f
        treb7uchet
    """.trimIndent().split("\n")

    println("Test Data: $test_data1")
    val tdA = part1(test_data1)
    println("Test 1: $tdA")
    require(tdA == 142) { "Test 1 failed: $tdA is not equal to 142" }

    val test_data2 = """
        two1nine
        eightwothree
        abcone2threexyz
        xtwone3four
        4nineeightseven2
        zoneight234
        7pqrstsixteen
    """.trimIndent().split("\n")

    println("Test Data: $test_data2")
    val tdB = part2(test_data2)
    println("Test 2: $tdB")
    require(tdB == 281) { "Test 2 failed: $tdB is not equal to 281" }

    val input = readInputData(INPUT_FILE)
    val part1_answer = part1(input)
    println("Part 1: $part1_answer")
    val part2_answer = part2(input)
    println("Part 2: $part2_answer")

}

File: 2023/1/src/Day01.py
from aocd.models import Puzzle
import re
import os
import time

# Constants
INPUT_FILE = "../input.txt"


def read_input_data(input_file):
    try:
        cwd = os.path.dirname(os.path.realpath(__file__))
        input_file = f"{cwd}/{input_file}"
        with open(input_file, "r") as file:
            data = file.read().split("\n")
        print(f"Using {input_file} for input data.")
    except FileNotFoundError:
        print(f"Input file {input_file} not found. Fetching data from AoC website.")
        puzzle = Puzzle(year=2023, day=6)
        data = puzzle.input_data.split("\n")
    return data


word_to_num = {
    "one": 1,
    "two": 2,
    "three": 3,
    "four": 4,
    "five": 5,
    "six": 6,
    "seven": 7,
    "eight": 8,
    "nine": 9,
}


def part1(input):
    total = 0
    # 1. Find first and last digit of each number
    # 2. Sum all numbers
    # 3. Return the sum

    for x in range(0, len(input)):
        digits = [ch for ch in input[x] if ch.isdigit()]
        # print(digits)
        if digits:  # Check if digits is not empty
            total += int(digits[0] + digits[-1])
    return total


def check_digits(input):
    val = ""
    if input in word_to_num:
        val = word_to_num[input]
        # print(f"Found {input} = {val}")
        return str(val)
    return str(input)


def part2(lines):
    total = 0
    spliter = "(?=(" + "|".join(word_to_num.keys()) + "|\\d))"
    for line in lines:
        digits = [*map(check_digits, re.findall(spliter, line))]
        if digits:
            # print(f"Digits: {digits}")
            total += int(digits[0] + digits[-1])
    return total

test_data1 = """
1abc2
pqr3stu8vwx
a1b2c3d4e5f
treb7uchet
"""

test_data2 = """
two1nine
eightwothree
abcone2threexyz
xtwone3four
4nineeightseven2
zoneight234
7pqrstsixteen
"""

if __name__ == "__main__":
    t_data = test_data1.split("\n")
    print(t_data)
    part1_answer = part1(t_data)
    print(f"Test 1: {part1_answer}")
    assert part1_answer == 142
    
    t_data = test_data2.split("\n")
    part2_answer = part2(t_data)
    print(f"Test 2: {part2_answer}")
    assert part2_answer == 281

    data = read_input_data(INPUT_FILE)

    start_p1 = time.perf_counter()
    part1_answer = part1(data)
    print(f"Part 1 answer: {part1_answer}")
    # puzzle.answer_a = part1_answer
    print(f"Part 1 time: {time.perf_counter() - start_p1:.5f} Sec.")

    start_p2 = time.perf_counter()
    part2_answer = part2(data)
    print(f"Part 2 answer: {part2_answer}")
    print(f"Part 2 time: {time.perf_counter() - start_p2} Sec.")
    # puzzle.answer_b = part2_answer
   
